nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java

```diff
@@ -0,0 +1,367 @@
+     * @param configurationContext at the time of configuration
+     * @throws AuthorizerCreationException for any issues configuring the provider
+     */
+    @Override
+    public void onConfigured(AuthorizerConfigurationContext configurationContext) throws AuthorizerCreationException {
+        // Our first init step is to select the command set based on the
+        // operating system name:
+        final String osName = System.getProperty("os.name");
+        ShellCommandsProvider commands = getCommandsProvider();
+
+        if (commands == null) {
+            if (osName.startsWith("Linux")) {
+                logger.debug("Selected Linux command set.");
+                commands = new NssShellCommands();
+            } else if (osName.startsWith("Mac OS X")) {
+                logger.debug("Selected OSX command set.");
+                commands = new OsxShellCommands();
+            } else {
+                throw new AuthorizerCreationException(OS_TYPE_ERROR);
+            }
+            setCommandsProvider(commands);
+        }
+
+        // Our second init step is to run the system check from that
+        // command set to determine if the other commands will work on
+        // this host or not.
+        try {
+            ShellRunner.runShell(commands.getSystemCheck());
+        } catch (final IOException ioexc) {
+            logger.error("initialize exception: " + ioexc + " system check command: " + commands.getSystemCheck());
+            throw new AuthorizerCreationException(SYS_CHECK_ERROR, ioexc.getCause());
+        }
+
+        // With our command set selected, and our system check passed,
+        // we can pull in the users and groups:
+        refreshUsers();
+        refreshGroups();
+
+        // Our last init step is to fire off the refresh threads:
+        Integer initialDelay = 30, fixedDelay = 30;
+        scheduler.scheduleWithFixedDelay(this::refreshUsers, initialDelay, fixedDelay, TimeUnit.SECONDS);
+        scheduler.scheduleWithFixedDelay(this::refreshGroups, initialDelay, fixedDelay, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Called immediately before instance destruction for implementers to release resources.
+     *
+     * @throws AuthorizerDestructionException If pre-destruction fails.
+     */
+    @Override
+    public void preDestruction() throws AuthorizerDestructionException {
+        try {
+            scheduler.shutdownNow();
+        } catch (final Exception ignored) {
+        }
+    }
+
+    public ShellCommandsProvider getCommandsProvider() {
+        return selectedShellCommands;
+    }
+
+    public void setCommandsProvider(ShellCommandsProvider commandsProvider) {
+        selectedShellCommands = commandsProvider;
+    }
+
+    private void refreshUsers() {
+        Map<String, User> byId = new HashMap<>();
+        Map<String, User> byName = new HashMap<>();
+        List<String> lines;
+
+        try {
+            lines = ShellRunner.runShell(selectedShellCommands.getUsersList());
```

Comment:
I'd like to see this added if it's helpful to someone, I'm just not sure how to best implement the feature.

Straight line count probably won't work because the same number of users could be added + deleted during the refresh interval, yielding different lists but the same line count.

We might be able to hash the command output + compare across runs; that would save the steps in rebuilding the objects.  On the other hand, we're not explicitly sorting, so we'd have to add that (e.g., embedded ` | sort `) to stabilize the record order.

My preference would be to re-visit this behavior later after we've got more experience with using the provider.

Reply:
I agree with what you just wrote. Sorting + storing/hashing would be sufficient but this change can wait as it is just a performance improvement.